### ISSUE TITLE
Proxy architecture

### ISSUE THREAD

**Issue by mmalecki** [npm/npm-registry-couchapp#128]
Since there's no good place to discuss this, I'm going to put this here.

Right now the npm registry is running behind a custom (closed-source) proxy written in Erlang, by @jhs. I'm not sure if at this point anyone else knows how to work this proxy, not to mention changing its functionality or fixing bugs. I feel like this is no longer acceptable, given how many people rely on npm these days.

I'd like to advocate switching to HAProxy for npm's hosting needs.

First of all, this is code someone wrote and, what's more important, [_tested_ for us](http://haproxy.1wt.eu/they-use-it.html). This also means significantly lower maintenance costs in the long run.

Second thing, it's **open source**. We can publish configs, have community poke at it. I feel like it'd be beneficial for our whole community _and_ people running it, if the code behind npm was open.

HAProxy is also **easy to run**. Configuration syntax is really simple. More people would be able to operate the infrastructure. There are tools around it (like [`bender`](https://github.com/vigour-io/bender) or [`thalassa`](https://github.com/PearsonEducation/thalassa)), which make running it in production a breeze by providing automatic configuration and service registry features.

It's also incredibly **safe**. For example, config reload doesn't shut down existing connections. Instead, it lets older workers finish and spawns new workers with reloaded config.

HAProxy also provides **automated failover** with retries and health checks of backend instances. 

To sum it up, I had nothing but great experience running HAProxy, both in front of npmjs.eu and node software. **I'd also be extremely happy to help with migration from the current Erlang proxy (including testing in staging, writing a load testing tool, etc.).** 

cc @isaacs @jhs
